chain/network/src/raw/connection.rs

```diff
@@ -136,41 +255,90 @@ impl Connection {
         Ok(peer)
     }
 
+    async fn on_accept(
+        stream: tcp::Stream,
+        secret_key: SecretKey,
+        chain_id: &str,
+        genesis_hash: CryptoHash,
+        head_height: BlockHeight,
+        tracked_shards: Vec<ShardId>,
+        archival: bool,
+        recv_timeout: Duration,
+    ) -> Result<Self, ConnectError> {
+        let mut stream = PeerStream::new(stream, recv_timeout);
+        let mut borsh_message_expected = true;
+        let (message, _timestamp) = match stream.recv_message().await {
+            Ok(m) => m,
+            Err(RecvError::Parse(len)) => {
+                tracing::debug!(target: "network", "dropping a non protobuf message of length {}. Probably an extra handshake.", len);
+                borsh_message_expected = false;
+                stream.recv_message().await?
+            }
+            Err(RecvError::IO(e)) => return Err(ConnectError::IO(e)),
+        };
+
+        let (peer_id, nonce) = match message {
+            // TODO: maybe check the handshake for sanity
+            PeerMessage::Tier2Handshake(h) => (h.sender_peer_id, h.partial_edge_info.nonce),
+            PeerMessage::HandshakeFailure(_peer_info, reason) => {
+                return Err(ConnectError::HandshakeFailure(reason))
+            }
+            _ => return Err(ConnectError::UnexpectedFirstMessage(message)),
+        };
+
+        let my_peer_id = PeerId::new(secret_key.public_key());
+        let handshake = new_handshake(
+            &secret_key,
+            &my_peer_id,
+            &peer_id,
+            stream.stream.local_addr.port(),
+            nonce,
+            PROTOCOL_VERSION,
+            chain_id,
+            genesis_hash,
+            head_height,
+            tracked_shards,
+            archival,
+        );
+
+        stream.write_message(&handshake).await.map_err(ConnectError::IO)?;
+
+        Ok(Self {
+            secret_key,
+            my_peer_id,
+            stream,
+            peer_id,
+            route_cache: lru::LruCache::new(1_000_000),
+            borsh_message_expected,
+        })
+    }
+
     async fn do_handshake(
         &mut self,
         protocol_version: ProtocolVersion,
         chain_id: &str,
         genesis_hash: CryptoHash,
         head_height: BlockHeight,
     ) -> Result<(), ConnectError> {
-        let handshake = PeerMessage::Tier2Handshake(Handshake {
+        let handshake = new_handshake(
+            &self.secret_key,
+            &self.my_peer_id,
+            &self.peer_id,
+            self.stream.stream.local_addr.port(),
+            1,
             protocol_version,
-            oldest_supported_version: protocol_version - 2,
-            sender_peer_id: self.my_peer_id.clone(),
-            target_peer_id: self.peer_id.clone(),
-            // we have to set this even if we have no intention of listening since otherwise
-            // the peer will drop our connection
-            sender_listen_port: Some(24567),
```

Comment:
Huh, this was hardcoded in the old code??

Reply:
Yeah, so the old code hardcoded a lot of the handshake. This `raw::Connection` is just used in the `ping` command (and `state-parts`), where we connect to a node and only send it pings. So much of the handshake just didn't matter since we had no intention of really participating in the protocol other than sending ping messages, and the next time we connect it'll be with a newly generated node key anyway. But yeah it was probably kinda sloppy to just hardcode it :D